File: algorithms/leetcode/208-implement-prefix-tree-trie/implement-prefix-tree-trie.js

class Trie {
    /**
     * Constructor
     */
    constructor() {
        this.tree = {};
    }
    
    /**
     * Inserts a word into the trie. 
     * @param {string} word
     * @return {void}
     */
    insert(word) {
        let pointer = this.tree;
        
        for (let i = 0; i < word.length; i++) {
            if (!pointer[word[i]]) {
                pointer[word[i]] = {};
            }
            
            pointer = pointer[word[i]];
        }
        
        pointer['$'] = {};
    }
    
    /**
     * Returns if the word is in the trie. 
     * @param {string} word
     * @return {boolean}
     */
    search(word) {
        let pointer = this.tree;
        
        for (let i = 0; i < word.length; i++) {
            if (!pointer[word[i]]) {
                return false;
            }
            
            pointer = pointer[word[i]];
        }
        
        return !!pointer['$'];
    }
    
    
    /**
     * Returns if there is any word in the trie that starts with the given prefix. 
     * @param {string} prefix
     * @return {boolean}
     */
    startsWith(prefix) {
        let pointer = this.tree;
        
        for (let i = 0; i < prefix.length; i++) {
            if (!pointer[prefix[i]]) {
                return false;
            }
            
            pointer = pointer[prefix[i]];
        }
        
        return true;
    }
}

File: algorithms/leetcode/208-implement-prefix-tree-trie/ImplementPrefixTreeTrie.java
class Trie {
    class TrieNode {
        TrieNode[] nodes;
        
        boolean isWord = false;
        
        TrieNode() {
            nodes = new TrieNode[26];
        }
    }
    
    TrieNode root;

    /** Initialize your data structure here. */
    public Trie() {
        root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.length(); i++) {
            int index = word.charAt(i) - 'a';
            if (node.nodes[index] == null) {
                node.nodes[index] = new TrieNode();
            }
            node = node.nodes[index];
        }
        
        node.isWord = true;
    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode node = root;
        for (int i = 0; i < word.length(); i++) {
            int index = word.charAt(i) - 'a';
            if (node.nodes[index] == null) {
                return false;
            }
            node = node.nodes[index];
        }
        
        return node.isWord;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode node = root;
        for (int i = 0; i < prefix.length(); i++) {
            int index = prefix.charAt(i) - 'a';
            if (node.nodes[index] == null) {
                return false;
            }
            node = node.nodes[index];
        }
        
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
